File: java-webflux-r2dbc/src/main/java/com/github/anddd7/api/ReactorController.java
package com.github.anddd7.api;

import com.github.anddd7.entity.Product;
import com.github.anddd7.entity.ProductStockDTO;
import java.math.BigDecimal;
import java.time.Duration;
import java.util.Random;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@RestController
@RequestMapping("/reactor")
public class ReactorController {

  private final ProductRepository productRepository;

  private WebClient webClient = WebClient.builder().baseUrl("http://localhost:18081").build();

  @Autowired
  public ReactorController(ProductRepository productRepository) {
    this.productRepository = productRepository;
  }

  @GetMapping("/product/{id}")
  public Mono<ProductStockDTO> findOne(@PathVariable Integer id) {
    Mono<Product> productMono = productRepository.findById(id);
    Mono<BigDecimal> stockMono = fetchStock(id);
    return productMono.zipWith(stockMono, ProductStockDTO::new);
  }

  private Mono<BigDecimal> fetchStock(Integer id) {
    // simulate unstable network
    int sleepTime = new Random(System.currentTimeMillis()).nextInt(1000);

    return webClient
        .get()
        .uri("/reactor/" + id + "/stock")
        .accept(MediaType.APPLICATION_JSON)
        .retrieve()
        .bodyToMono(BigDecimal.class)
        .delayElement(Duration.ofMillis(Integer.valueOf(sleepTime).longValue()));
  }

  @GetMapping("/product")
  public Flux<ProductStockDTO> findAll() {
    return productRepository.findAll().flatMap(this::fetchProductStock);
  }

  private Mono<ProductStockDTO> fetchProductStock(Product product) {
    Mono<BigDecimal> stockMono = fetchStock(product.getId());
    return stockMono.map(stock -> new ProductStockDTO(product, stock));
  }


  @GetMapping("/{id}/stock")
  public BigDecimal stock(@PathVariable Integer id) {
    // mock 3rd api, return immediately
    Random r = new Random(System.currentTimeMillis());
    double positiveDouble = Math.abs(r.nextDouble());
    return BigDecimal.valueOf(positiveDouble);
  }
}

File: kotlin-webflux-r2dbc/src/main/kotlin/com/github/anddd7/api/ReactorController.kt
package com.github.anddd7.api

import com.github.anddd7.dto.Product
import com.github.anddd7.dto.ProductStockDTO
import org.springframework.data.repository.reactive.ReactiveCrudRepository
import org.springframework.http.MediaType
import org.springframework.stereotype.Repository
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.bodyToMono
import reactor.core.publisher.Mono
import java.math.BigDecimal
import java.time.Duration
import kotlin.math.abs
import kotlin.random.Random

@RestController
@RequestMapping("/reactor")
class ReactorController(
    private val productRepository: ProductRepository
) {
  private val webClient = WebClient.builder().baseUrl("http://localhost:18084").build()

  @RequestMapping("/product/{id}")
  fun findOne(@PathVariable id: Int) =
      productRepository.findById(id).zipWith(fetchStock(id))
          .map { ProductStockDTO(it.t1, it.t2) }

  @RequestMapping("/product")
  fun findAll() =
      productRepository.findAll().flatMap { product ->
        fetchStock(product.id).map { stock -> ProductStockDTO(product, stock) }
      }

  private fun fetchStock(id: Int) =
      webClient
          .get()
          .uri("/reactor/$id/stock")
          .accept(MediaType.APPLICATION_JSON)
          .retrieve()
          .bodyToMono<BigDecimal>()

  @GetMapping("/{id}/stock")
  fun stock(@PathVariable id: Int) =
      Mono.fromCallable { Random(id).nextFloat().let(::abs).toBigDecimal() }
          .delayElement(Duration.ofMillis(Random(id).nextLong(1, 1000)))
}

@Repository
interface ProductRepository : ReactiveCrudRepository<Product, Int>
